File: src/editor/freehand/wobbleSmoother.ts
import { vec2 } from '@/math';
import { normalize01 } from './utils';

abstract class WobbleSmoother {
  private static m_samples: Sample[];
  private static m_weightedPositionSum: vec2 = [0, 0];
  private static m_distanceSum: number = 0;
  private static m_durationSum: number = 0;
  private static m_params: WobbleSmootherParams;

  static reset(params: WobbleSmootherParams, position: vec2, time: number): void {
    this.m_params = params;
    this.m_samples = [{ position, time, weightedPosition: [0, 0], distance: 0, duration: 0 }];
    this.m_weightedPositionSum = [0, 0];
    this.m_distanceSum = 0;
    this.m_durationSum = 0;
  }

  static update(position: vec2, time: number): vec2 {
    const back = this.m_samples[this.m_samples.length - 1];
    const deltaTime = time - back.time;

    this.m_samples.push({
      position,
      weightedPosition: vec2.mulS(position, deltaTime),
      distance: vec2.dist(position, back.position),
      duration: deltaTime,
      time
    });

    vec2.add(this.m_weightedPositionSum, back.weightedPosition, this.m_weightedPositionSum);
    this.m_distanceSum += back.distance;
    this.m_durationSum += back.duration;

    while (this.m_samples[0].time < time - this.m_params.timeout) {
      const front = this.m_samples[0];

      vec2.sub(this.m_weightedPositionSum, front.weightedPosition, this.m_weightedPositionSum);
      this.m_distanceSum -= front.distance;
      this.m_durationSum -= front.duration;

      this.m_samples.shift();
    }

    if (this.m_durationSum == 0) return position;

    const avgPosition = vec2.divS(this.m_weightedPositionSum, this.m_durationSum);
    const avgSpeed = this.m_distanceSum / this.m_durationSum;

    return vec2.lerp(
      avgPosition,
      position,
      normalize01(this.m_params.speedFloor, this.m_params.speedCeiling, avgSpeed)
    );
  }
}

export default WobbleSmoother;

File: wasm-src/editor/input/wobble_smoother.cpp
#include "wobble_smoother.h"

#include "../../math/vector.h"
#include "../../math/math.h"
#include "../../utils/console.h"

std::deque<WobbleSmoother::Sample> WobbleSmoother::s_samples{};
vec2 WobbleSmoother::s_weighted_position_sum = { 0.0f, 0.0f };
float WobbleSmoother::s_distance_sum = 0.0f;
float WobbleSmoother::s_weighted_pressure_sum = 0.0f;
double WobbleSmoother::s_duration_sum = 0.0;
WobbleSmoother::WobbleSmootherParams WobbleSmoother::s_params{};

void WobbleSmoother::reset(const WobbleSmootherParams& params, const vec2& position, float pressure, double time) {
  s_params = params;

  s_samples.clear();
  s_weighted_position_sum = { 0.0f, 0.0f };
  s_weighted_pressure_sum = 0.0f;
  s_distance_sum = 0.0f;
  s_duration_sum = 0.0;

  s_samples.push_back({ position, { 0.0f, 0.0f }, pressure, 0.0f, 0.0f, 0.0f, time });
}

void WobbleSmoother::reset(const vec2& position, float pressure, double time) {
  reset({}, position, pressure, time);
}

vec3 WobbleSmoother::update(const vec2& position, float pressure, double time) {
  double delta_time = time - s_samples.back().time;
  float dist = distance(position, s_samples.back().position);
  float prev_pressure = s_samples.back().pressure;

  Sample sample = {
    position,
    position * (float)delta_time,
    pressure,
    pressure * (float)delta_time,
    dist,
    delta_time,
    time
  };


  s_samples.push_back(sample);

  s_weighted_position_sum += sample.weighted_position;
  s_weighted_pressure_sum += sample.weighted_pressure;
  s_distance_sum += sample.distance;
  s_duration_sum += sample.duration;

  while (s_samples.front().time < time - s_params.timeout) {
    s_weighted_position_sum -= s_samples.front().weighted_position;
    s_weighted_pressure_sum -= s_samples.front().weighted_pressure;
    s_distance_sum -= s_samples.front().distance;
    s_duration_sum -= s_samples.front().duration;

    s_samples.pop_front();
  }

  if (s_duration_sum == 0.0f) {
    return { position.x, position.y, pressure };
  }

  vec2 average_position = s_weighted_position_sum / (float)s_duration_sum;
  float average_speed = s_distance_sum / (float)s_duration_sum;
  float average_pressure = s_weighted_pressure_sum / (float)s_duration_sum;

  float t = normalize(average_speed);
  vec2 value = lerp(average_position, position, t);

  if (s_params.simulate_pressure) {
    float sp = std::min(1.0f, dist / s_params.width);
    float rp = std::min(1.0f, 1.0f - sp);
    return { value.x, value.y, std::min(1.0f, prev_pressure + (rp - prev_pressure) * (sp * 0.675f)) };
  }

  return { value.x, value.y, lerp(average_pressure, pressure, t) };
}

float WobbleSmoother::normalize(float value) {
  if (s_params.speed_floor == s_params.speed_ceiling) {
    return value > s_params.speed_floor ? 1.0f : 0.0f;
  }

  return clamp((value - s_params.speed_floor) / (s_params.speed_ceiling - s_params.speed_floor), 0.0f, 1.0f);
}
